lib/Mojo/WebSocketProxy.pm

```diff
@@ -0,0 +1,252 @@
+                    ]
+                ],
+                base_path => '/api',
+                url => 'http://rpc-host.com:8080/',
+            }
+        );
+   }
+
+=head1 DESCRIPTION
+
+Using this module you can forward websocket JSON-RPC 2.0 requests to RPC server.
+
+For every message it creates separate hash ref storage, which is available from hooks as $req_storage.
+Request storage have RPC call parameters in $req_storage->{call_params}.
+It copies message args to $req_storage->{call_params}->{args}.
+You can use Mojolicious stash to store data between messages in one connection.
+
+=head1 Proxy responses
+
+The plugin sends websocket messages to clietn with RPC response data.
+If RPC reponse looks like this:
+    
+    {status => 1}
+    
+It returns simple response like this:
+    
+    {$msg_type => 1, msg_type => $msg_type}
+
+If RPC returns something like this:
+    
+    {
+        response_data => [..],
+        status        => 1,
+    }
+    
+Plugin returns common response like this:
+    
+    {
+        $msg_type => $rpc_response,
+        msg_type  => $msg_type,
+    }
+
+You can customize ws porxy response using 'response' hook.
+
+=head1 Plugin parameters
+
+The plugin understands the following parameters.
+
+=head2 actions
+
+A pointer to array of action details, which contain stash_params, 
+request-response callbacks, other call parameters.
+
+    $self->plugin(
+        'web_socket_proxy' => {
+            actions => [ 
+                ['action1_json_key', {details_key1 => details_value1}],
+                ['action2_json_key']
+            ]
+        });
+        
+=head2 before_forward
+
+    before_forward => [sub { my ($c, $req_storage) = @_; ... }, sub {...}]
+
+Global hooks which will run after request is dispatched and before to start preparing RPC call.
+It'll run every hook or until any hook returns some non-empty result.
+If returns any hash ref then that value will be JSON encoded and send to client,
+without forward action to RPC. To call RPC every hook should return empty or undefined value.
+It's good place to some validation or subscribe actions.
+        
+=head2 after_forward
+    
+    after_forward => [sub { my ($c, $result, $req_storage) = @_; ... }, sub {...}]
+
+Global hooks which will run after every forwarded RPC call done.
+Or even forward action isn't running.
+It can view or modify result value from 'before_forward' hook.
+It'll run every hook or until any hook returns some non-empty result.
+If returns any hash ref then that value will be JSON encoded and send to client.
+
+=head2 after_dispatch
+
+    after_dispatch => [sub { my $c = shift; ... }, sub {...}]
+    
+Global hooks which will run at the end of request handling.
+
+=head2 before_get_rpc_response (global)
+
+    before_get_rpc_response => [sub { my ($c, $req_storage) = @_; ... }, sub {...}]
+
+Global hooks which will run when asynchronous RPC call is answered.
+
+=head2 after_got_rpc_response (global)
```

Comment:
@tfoertsch @evostrov 

Maybe generalize before/after hooks to `around`? We'll get twice less functions?